lib/api/objectPutTagging.js

```diff
@@ -0,0 +1,95 @@
+import aclUtils from '../utilities/aclUtils';
+
+import async from 'async';
+import { errors } from 'arsenal';
+
+import { decodeVersionId, getVersionIdResHeader }
+  from './apiUtils/object/versioning';
+
+import { metadataValidateBucketAndObj } from '../metadata/metadataUtils';
+import { pushMetric } from '../utapi/utilities';
+import collectCorsHeaders from '../utilities/collectCorsHeaders';
+import metadata from '../metadata/wrapper';
+import { parseTagXml } from './apiUtils/object/tagging';
+
+/**
+ * Object Put Tagging - Adds tag(s) to object
+ * @param {AuthInfo} authInfo - Instance of AuthInfo class with requester's info
+ * @param {object} request - http request object
+ * @param {object} log - Werelogs logger
+ * @param {function} callback - callback to server
+ * @return {undefined}
+ */
+export default function objectPutTagging(authInfo, request, log, callback) {
+    log.debug('processing request', { method: 'objectPutTagging' });
+
+    if (!aclUtils.checkGrantHeaderValidity(request.headers)) {
+        log.trace('invalid acl header');
+        return callback(errors.InvalidArgument);
+    }
+
+    const bucketName = request.bucketName;
+    const objectKey = request.objectKey;
+
+    const decodedVidResult = decodeVersionId(request.query);
+    if (decodedVidResult instanceof Error) {
+        log.trace('invalid versionId query', {
+            versionId: request.query.versionId,
+            error: decodedVidResult,
+        });
+        return callback(decodedVidResult);
+    }
+    const reqVersionId = decodedVidResult;
+
+    const metadataValParams = {
+        authInfo,
+        bucketName,
+        objectKey,
+        requestType: 'bucketOwnerAction',
+        versionId: reqVersionId,
+    };
+
+    return async.waterfall([
+        next => metadataValidateBucketAndObj(metadataValParams, log,
+          (err, bucket, objectMD) => next(err, bucket, objectMD)),
+        (bucket, objectMD, next) => {
+            log.trace('parsing tag(s)');
+            return parseTagXml(request.post, log, (err, tags) => {
+                if (err) {
+                    return next(err);
+                }
+                if (!objectMD) {
+                    const err = reqVersionId ? errors.NoSuchVersion :
+                        errors.NoSuchKey;
+                    return next(err, bucket);
+                }
```

Comment:
probably, you may need to add one more check if the version/object you got is a delete marker. It seems the error in that case is usually MethodNotAllowed, or InvalidRequest, if the version was specified in the query, or NoSuchKey if no version was specified (just made a similar fix here for objAcl apis: https://github.com/scality/S3/pull/713/files).